File: core/units.py
"""
ballistics.ge - Unit System Formatters
Centralized formatters that respect ``st.session_state.units``
("metric" | "imperial"). Use these helpers anywhere user-facing
numbers are rendered so the toggle flips everything consistently.

Internally, the solver always works in metric. These helpers convert
only at display time.
"""
from typing import Literal

import streamlit as st

UnitSystem = Literal["metric", "imperial"]

_MPS_TO_FPS = 3.28084
_FPS_TO_MPS = 1 / _MPS_TO_FPS
_M_TO_YD = 1.09361
_YD_TO_M = 1 / _M_TO_YD
_KG_TO_LB = 2.20462
_MBAR_TO_INHG = 0.02953
_INHG_TO_MBAR = 1 / _MBAR_TO_INHG
_MPH_TO_MPS = 0.44704
_MPS_TO_MPH = 1 / _MPH_TO_MPS
_M_TO_FT = 3.28084
_FT_TO_M = 1 / _M_TO_FT


def current_system() -> UnitSystem:
    return st.session_state.get("units", "metric")


def is_imperial() -> bool:
    return current_system() == "imperial"


# --- Distance --------------------------------------------------------------

def fmt_range(meters: float, precision: int = 0) -> str:
    if is_imperial():
        return f"{meters * _M_TO_YD:.{precision}f} yd"
    return f"{meters:.{precision}f} m"


def fmt_distance_short(meters: float) -> str:
    """Short inline form without unit suffix — handy for table cells."""
    return f"{meters * _M_TO_YD:.0f}" if is_imperial() else f"{meters:.0f}"


def range_label() -> str:
    return "yd" if is_imperial() else "m"


# --- Velocity --------------------------------------------------------------

def fmt_velocity(mps: float, precision: int = 0) -> str:
    if is_imperial():
        return f"{mps * _MPS_TO_FPS:.{precision}f} fps"
    return f"{mps:.{precision}f} m/s"


def velocity_label() -> str:
    return "fps" if is_imperial() else "m/s"


# --- Temperature -----------------------------------------------------------

def fmt_temperature(celsius: float, precision: int = 0) -> str:
    if is_imperial():
        return f"{celsius * 9 / 5 + 32:.{precision}f} °F"
    return f"{celsius:.{precision}f} °C"


# --- Pressure --------------------------------------------------------------

def fmt_pressure(mbar: float) -> str:
    if is_imperial():
        return f"{mbar * _MBAR_TO_INHG:.2f} inHg"
    return f"{mbar:.0f} mbar"


# --- Energy ----------------------------------------------------------------

def fmt_energy(joules: float) -> str:
    if is_imperial():
        # 1 J = 0.737562 ft·lbf
        return f"{joules * 0.737562:.0f} ft·lb"
    return f"{joules:.0f} J"


# --- Input-side converters (imperial user-input → metric internal) ----------

def input_range_to_m(value: float) -> float:
    """Convert a range from the user's unit system to metres."""
    return value * _YD_TO_M if is_imperial() else value


def input_range_from_m(meters: float) -> float:
    """Convert metres to the user's display unit for pre-filling inputs."""
    return meters * _M_TO_YD if is_imperial() else meters


def input_speed_to_mps(value: float) -> float:
    """Convert wind speed from user's unit (mph/m/s) to m/s."""
    return value * _MPH_TO_MPS if is_imperial() else value


def input_speed_from_mps(mps: float) -> float:
    return mps * _MPS_TO_MPH if is_imperial() else mps


def input_temp_to_c(value: float) -> float:
    return (value - 32) * 5 / 9 if is_imperial() else value


def input_temp_from_c(celsius: float) -> float:
    return celsius * 9 / 5 + 32 if is_imperial() else celsius


def input_pressure_to_mbar(value: float) -> float:
    return value * _INHG_TO_MBAR if is_imperial() else value


def input_pressure_from_mbar(mbar: float) -> float:
    return mbar * _MBAR_TO_INHG if is_imperial() else mbar


def input_alt_to_m(value: float) -> float:
    return value * _FT_TO_M if is_imperial() else value


def input_alt_from_m(meters: float) -> float:
    return meters * _M_TO_FT if is_imperial() else meters


def speed_label() -> str:
    return "mph" if is_imperial() else "m/s"


def temp_label() -> str:
    return "°F" if is_imperial() else "°C"


def pressure_label() -> str:
    return "inHg" if is_imperial() else "mbar"


def alt_label() -> str:
    return "ft" if is_imperial() else "m"


# --- Angular units & scope clicks -----------------------------------------

MRAD_TO_MOA = 3.43775

# Label -> click size in MRAD
CLICK_OPTIONS = {
    "0.1 MRAD": 0.1,
    "0.05 MRAD": 0.05,
    "1/4 MOA": 0.25 / MRAD_TO_MOA,
    "1/8 MOA": 0.125 / MRAD_TO_MOA,
    "1/2 MOA": 0.5 / MRAD_TO_MOA,
}


def angular_unit() -> str:
    """'MRAD' or 'MOA' as chosen in Settings."""
    return st.session_state.get("angular_unit", "MRAD")


def click_value_mrad() -> float:
    label = st.session_state.get("click_value", "0.1 MRAD")
    return CLICK_OPTIONS.get(label, 0.1)


def to_angular(mrad: float) -> float:
    return mrad * MRAD_TO_MOA if angular_unit() == "MOA" else mrad


def fmt_angular(mrad: float, precision: int = 2, signed: bool = False) -> str:
    val = to_angular(mrad)
    spec = f"{{:{'+' if signed else ''}.{precision}f}}"
    return f"{spec.format(val)} {angular_unit()}"


def clicks_for(mrad: float) -> int:
    """Number of scope clicks (rounded to nearest) for an angular value."""
    return int(round(abs(mrad) / click_value_mrad()))


# --- Linear drop at target -----------------------------------------------

def fmt_drop_linear(meters: float) -> str:
    if is_imperial():
        return f"{meters * 39.3701:.1f} in"
    return f"{meters * 100:.1f} cm"


# --- Small-length inputs (sight height) ----------------------------------

def sight_height_label() -> str:
    return "in" if is_imperial() else "mm"


def input_sight_height_to_mm(value: float) -> float:
    return value * 25.4 if is_imperial() else value


def input_sight_height_from_mm(mm: float) -> float:
    return mm / 25.4 if is_imperial() else mm


def input_velocity_to_mps(value: float) -> float:
    return value * _FPS_TO_MPS if is_imperial() else value


def input_velocity_from_mps(mps: float) -> float:
    return mps * _MPS_TO_FPS if is_imperial() else mps


def roundtrip(stored_metric: float, shown: float, entered: float, to_metric) -> float:
    """Keep the stored metric value unless the user actually changed the field.

    Display values are rounded (e.g. 850 m/s → 2789 fps), so converting the
    unchanged display value back would slowly drift the stored number
    (850 → 850.09). Only convert when the widget value differs from what
    was seeded into it.
    """
    if abs(float(entered) - float(shown)) < 1e-9:
        return float(stored_metric)
    return float(to_metric(entered))

File: web/src/lib/units.ts
/** Unit helpers. The solver is metric; conversion happens at the UI edge. */
export type UnitSystem = "metric" | "imperial";
export type Angular = "MRAD" | "MOA";

export const MRAD_TO_MOA = 3.43775;
export const M_TO_YD = 1.09361;
export const MPS_TO_FPS = 3.28084;
export const MBAR_TO_INHG = 0.02953;
export const MPH_TO_MPS = 0.44704;

export const CLICK_OPTIONS: Record<string, number> = {
  "0.1 MRAD": 0.1,
  "0.05 MRAD": 0.05,
  "1 cm/100 m": 0.1,
  "1/4 MOA": 0.25 / MRAD_TO_MOA,
  "1/8 MOA": 0.125 / MRAD_TO_MOA,
  "1/3 MOA": (1 / 3) / MRAD_TO_MOA,
  "1/2 MOA": 0.5 / MRAD_TO_MOA,
};

export const fmt = (v: number, d = 1) => v.toFixed(d);

export function rangeLabel(u: UnitSystem) { return u === "imperial" ? "yd" : "m"; }
export function speedLabel(u: UnitSystem) { return u === "imperial" ? "mph" : "m/s"; }
export function velLabel(u: UnitSystem) { return u === "imperial" ? "fps" : "m/s"; }
export function tempLabel(u: UnitSystem) { return u === "imperial" ? "°F" : "°C"; }
export function pressLabel(u: UnitSystem) { return u === "imperial" ? "inHg" : "mbar"; }
export function altLabel(u: UnitSystem) { return u === "imperial" ? "ft" : "m"; }
export function smallLabel(u: UnitSystem) { return u === "imperial" ? "in" : "cm"; }
export function sightLabel(u: UnitSystem) { return u === "imperial" ? "in" : "mm"; }

export const rangeFrom = (m: number, u: UnitSystem) => (u === "imperial" ? m * M_TO_YD : m);
export const rangeTo = (v: number, u: UnitSystem) => (u === "imperial" ? v / M_TO_YD : v);
export const velFrom = (mps: number, u: UnitSystem) => (u === "imperial" ? mps * MPS_TO_FPS : mps);
export const velTo = (v: number, u: UnitSystem) => (u === "imperial" ? v / MPS_TO_FPS : v);
export const windFrom = (mps: number, u: UnitSystem) => (u === "imperial" ? mps / MPH_TO_MPS : mps);
export const windTo = (v: number, u: UnitSystem) => (u === "imperial" ? v * MPH_TO_MPS : v);
export const tempFrom = (c: number, u: UnitSystem) => (u === "imperial" ? c * 9 / 5 + 32 : c);
export const tempTo = (v: number, u: UnitSystem) => (u === "imperial" ? (v - 32) * 5 / 9 : v);
export const pressFrom = (mb: number, u: UnitSystem) => (u === "imperial" ? mb * MBAR_TO_INHG : mb);
export const pressTo = (v: number, u: UnitSystem) => (u === "imperial" ? v / MBAR_TO_INHG : v);
export const altFrom = (m: number, u: UnitSystem) => (u === "imperial" ? m * MPS_TO_FPS : m);
export const altTo = (v: number, u: UnitSystem) => (u === "imperial" ? v / MPS_TO_FPS : v);
export const smallFrom = (cm: number, u: UnitSystem) => (u === "imperial" ? cm / 2.54 : cm);
export const smallTo = (v: number, u: UnitSystem) => (u === "imperial" ? v * 2.54 : v);
export const sightFrom = (mm: number, u: UnitSystem) => (u === "imperial" ? mm / 25.4 : mm);
export const sightTo = (v: number, u: UnitSystem) => (u === "imperial" ? v * 25.4 : v);

export const toAngular = (mrad: number, a: Angular) => (a === "MOA" ? mrad * MRAD_TO_MOA : mrad);
export const fromAngular = (v: number, a: Angular) => (a === "MOA" ? v / MRAD_TO_MOA : v);
export const clicksFor = (mrad: number, clickMrad: number) => Math.round(Math.abs(mrad) / clickMrad);

export function fmtRange(m: number, u: UnitSystem) { return `${Math.round(rangeFrom(m, u))} ${rangeLabel(u)}`; }
export function fmtVel(mps: number, u: UnitSystem, d = 0) { return `${velFrom(mps, u).toFixed(d)} ${velLabel(u)}`; }
export function fmtTemp(c: number, u: UnitSystem) { return `${tempFrom(c, u).toFixed(0)} ${tempLabel(u)}`; }
export function fmtPress(mb: number, u: UnitSystem) { return u === "imperial" ? `${pressFrom(mb, u).toFixed(2)} inHg` : `${mb.toFixed(0)} mbar`; }
export function fmtDrop(m: number, u: UnitSystem) { return u === "imperial" ? `${(m * 39.3701).toFixed(1)} in` : `${(m * 100).toFixed(1)} cm`; }
export function fmtEnergy(j: number, u: UnitSystem) { return u === "imperial" ? `${(j * 0.737562).toFixed(0)} ft·lb` : `${j.toFixed(0)} J`; }
export function fmtAng(mrad: number, a: Angular, d = 2, signed = false) {
  const v = toAngular(mrad, a);
  return `${signed && v > 0 ? "+" : ""}${v.toFixed(d)} ${a}`;
}
